### DISCUSSION TITLE
New executable for NOAA gfs_wave_post workflow

### DISCUSSION THREAD

**Discussion by dkokron** [NOAA-EMC/WW3#1304] [Ideas]
I modified the existing ww3_outp.ftn code and build process to create a new executable for use in the NOAA gfs_wave_post workflow.  Instead of parallelizing the report generation and taking 768 processes (three nodes) over 15 minutes, the new executable (ww3_outp2) does the job in under three minutes using a single core.
What is the process for getting these changes into WW3 so that NOAA can use them?

**Comment by MatthewMasarik-NOAA** [comment]
Hi @dkokron.  Great, that's sounds like a massive improvement.  This would involve committing to a branch that is already in operations, and I don't know what that full process entails. @JessicaMeixner-NOAA are you able to advise?

**Comment by JessicaMeixner-NOAA** [comment]
@dkokron  That's amazing!   Can you start by sharing your WW3 code and your g-w updates with us for testing on our end?

**Reply by JessicaMeixner-NOAA** [reply]
FYI @AliS-Noaa as you were working on similar things

**Reply by dkokron** [reply]
What does g-w stand for?

**Reply by JessicaMeixner-NOAA** [reply]
g-w is global-workflow: https://github.com/noAA-EMC/global-workflow

**Comment by dkokron** [comment]
Does you team have access to acorn?

**Reply by JessicaMeixner-NOAA** [reply]
Yes, I have access to acorn, but I prefer if you can share code via github.

**Comment by dkokron** [comment]
I'm a rusty on git collaboration.  Here is my understanding.  I need to fork WW3 in github, clone my fork to acorn, branch master/main develop, update my branch, push my branch back to my fork then submit a PR.  Did I get that right?

**Reply by JessicaMeixner-NOAA** [reply]
Yes, I'm assuming the code that you started with is from this branch though: https://github.com/NOAA-EMC/WW3/tree/production/GFS.v16 and not the main/develop branch.  If you just want to push it to a branch we can start with that before we make a PR.    We would want to test and then also do that fairly close to having an implementation.   

I'm currently on a detail, so my time is limited but if you'd prefer I can set up a meting for someone to help walk through things and if no one else is available I can do that myself.

**Reply by dkokron** [reply]
git reflog shows the following
6dc6c39e (HEAD, origin/production/GFS.v16) HEAD@{0}: checkout: moving from develop to 6dc6c39e9deb49dbe2b9515eb70f8d7571989a3f
2bf204eb (origin/develop, origin/HEAD, develop) HEAD@{1}: clone: from https://github.com/NOAA-EMC/WW3

**Reply by dkokron** [reply]
for the record, my sandbox is on acorn under /lfs/h1/hpc/support/daniel.kokron/Projects/gfs_wave_post_bndpntbll/gfs.v16.3.8/sorc/fv3gfs.fd/WW3

**Reply by JessicaMeixner-NOAA** [reply]
I don't see that any files have changes?  I don't know if it's a permissions issue or something else?

**Reply by dkokron** [reply]
I'm in the middle of cloning stuff.  Look in the  WW3.orig directory

**Reply by dkokron** [reply]
Here is the list of changes under WW3.orig.  I'll port these to my fork/clone then test using the case I have on acorn.

HEAD detached at 6dc6c39e
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   model/bin/make_makefile.sh
	modified:   model/bin/w3_automake
	modified:   model/bin/w3_make
	modified:   model/bin/w3_source
	modified:   model/esmf/Makefile
	modified:   model/ftn/w3bullmd.ftn
	modified:   model/ftn/w3iogrmd.ftn
	modified:   model/ftn/w3iopomd.ftn

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	WW3_INSTALL/
	exec/
	model/esmf/build.sh
	model/esmf/gmake.log
	model/ftn/ww3_outp.ftn.dkokron
	model/ftn/ww3_outp2.ftn

**Reply by dkokron** [reply]
Should I start from branch production/GFS.v16?

**Reply by JessicaMeixner-NOAA** [reply]
> Should I start from branch production/GFS.v16?

If our goal is to get this into v16, then yes.  We'll have to update and port things to the new code for v17, but first we can just look at things for v16 as I'm assuming the idea is to get this fix in sooner than later.

**Comment by dkokron** [comment]
I just tested this branch using the gfs_wave_post case.  Everything appears to be fine.
https://github.com/dkokron/WW3/tree/dkokron/gfs_wave_post

Let me know if I need to change anything.

**Comment by dkokron** [comment]
The new ww3_outp2 executable takes one command line argument which is the prefix for the output files (currently "gfswave").  The scripting creates the ww3_outp.inp input file on-the-fly.  You can still use the original scripting and executable by changing line 348 of /lfs/h1/hpc/support/daniel.kokron/Projects/gfs_wave_post_bndpntbll/gfs.v16.3.8/scripts/exgfs_wave_post_pnt.sh
from
  if true; then
to
  if false; then

**Comment by dkokron** [comment]
I have a bug in my exgfs_wave_post_pnt.sh.  I don't expect fixing it will change the overall approach.  I'll let you know when it's fixed.

**Comment by dkokron** [comment]
Just pushed some changes to my github branch.  I also modified my copy of gfs.v16.3.8/scripts/exgfs_wave_post_pnt.sh on acorn.  I tested modifications using the case I have on-hand.  I found no differences.

**Comment by dkokron** [comment]
I will be on PTO, without access to NOAA systems, starting tonight and ending 12 Oct.

**Reply by JessicaMeixner-NOAA** [reply]
Thanks for letting us know.  Hopefully we'll have had time to review and test everything on our end by the time you return from leave.  Hope you enjoy your time away!

**Comment by dkokron** [comment]
Has there been any movement on this discussion?

**Reply by JessicaMeixner-NOAA** [reply]
No apologies - I have not had a chance to look at this and we did not have other available resources from others on the team.

**Reply by dkokron** [reply]
Please let me know if there is any further testing needed.

FYI, the new executable has limited functionality compared with the ww3_outp it is based on.  The other functionality is not tested and likely broken in the new executable.

**Reply by JessicaMeixner-NOAA** [reply]
Thanks for the heads up on this.  We would like to test this on our end as well to investigate further but have not quite got a chance.
